wrench/src/reftest.rs

```diff
@@ -93,108 +93,144 @@ impl ReftestImage {
-                runner(Reftest {
-                    op: kind,
-                    test: test.as_path(),
-                    reference: reference.as_path(),
-                });
-            }
-            _ => panic!(),
-        };
+        ReftestManifest {
+            reftests: reftests
+        }
     }
 
+    fn find(&self, path: &Path) -> Option<&Reftest> {
+        self.reftests.iter().find(|x| x.test == path || x.reference == path)
+    }
 }
 
-fn render_yaml(wrench: &mut Wrench,
-               window: &mut WindowWrapper,
-               filename: &Path,
-               rx: &Receiver<()>)
-               -> ReftestImage {
-    let mut reader = YamlFrameReader::new(filename);
-    reader.do_frame(wrench);
-    // wait for the frame
-    rx.recv().unwrap();
-    wrench.render();
-
-    let size = window.get_inner_size();
-    let pixels = gl::read_pixels(0,
-                                 0,
-                                 size.0 as gl::GLsizei,
-                                 size.1 as gl::GLsizei,
-                                 gl::RGBA,
-                                 gl::UNSIGNED_BYTE);
-    window.swap_buffers();
-
-    ReftestImage {
-        data: pixels,
-        size: DeviceUintSize::new(size.0, size.1)
-    }
+pub struct ReftestHarness<'a> {
+    wrench: &'a mut Wrench,
+    window: &'a mut WindowWrapper,
+    rx: Receiver<()>,
 }
+impl<'a> ReftestHarness<'a> {
+    pub fn new(wrench: &'a mut Wrench,
+               window: &'a mut WindowWrapper) -> ReftestHarness<'a>
+    {
+        // setup a notifier so we can wait for frames to be finished
+        struct Notifier {
+            tx: Sender<()>,
+        };
+        impl RenderNotifier for Notifier {
+            fn new_frame_ready(&mut self) {
+                self.tx.send(()).unwrap();
+            }
+            fn new_scroll_frame_ready(&mut self, _composite_needed: bool) {}
+        }
+        let (tx, rx) = channel();
+        wrench.renderer.set_render_notifier(Box::new(Notifier { tx: tx }));
 
-pub fn run_reftests(wrench: &mut Wrench, window: &mut WindowWrapper, filename: &str) {
-    // setup a notifier so we can wait for frames to be finished
-    struct Notifier {
-        tx: Sender<()>,
-    };
-    impl RenderNotifier for Notifier {
-        fn new_frame_ready(&mut self) {
-            self.tx.send(()).unwrap();
+        ReftestHarness {
+            wrench: wrench,
+            window: window,
+            rx: rx,
         }
-        fn new_scroll_frame_ready(&mut self, _composite_needed: bool) {}
     }
-    let (tx, rx) = channel();
-    wrench.renderer.set_render_notifier(Box::new(Notifier { tx: tx }));
 
-    let mut total_passing = 0;
-    let mut total_failing = 0;
+    pub fn run(mut self, base_manifest: &Path, specific_reftest: Option<&Path>) {
+        let manifest = ReftestManifest::new(base_manifest);
+        let mut total_passing = 0;
+        let mut total_failing = 0;
+
+        match specific_reftest {
```

Comment:
I had a hard time fighting with the type checker on this with the incompatability of [&Reftest] and [Reftest]. I restructed the code to instead always search for a prefix match of the input. So you can do `wrench reftest reftests/blend` for all the reftests in that folder. This makes both cases the same.